### ISSUE TITLE
OPENMVG_STD_UNORDERED_MAP not passed to targets compile options anymore

### ISSUE THREAD

**Issue by sebastienchappuis** [openMVG/openMVG#1417]
Hello !

I noticed that since 1.4, `OPENMVG_STD_UNORDERED_MAP` is not passed as a compile option anymore to the installed targets, which breaks things if you build your project with **clang**, install it, and import it from another library.

From my understanding, this is because [`register_definitions`](https://github.com/openMVG/openMVG/blob/master/src/CMakeLists.txt#L85) is not used the same way it used to be in 1.3. 

I noticed this `OpenMVG_CFLAGS` variable [here](https://github.com/openMVG/openMVG/blob/master/src/CMakeLists.txt#L552-L556) which seems to tackle that problem but does not.

I guess the "modern" cmake solution would be to use `target_compile_options()` on each lib that uses `types.hpp` (which is almost every lib if I'm correct).

Do you think there is a better solution ?

Cheers

**Comment by pmoulon** [comment]
Thx for noticing this.

`target_compile_options` is the correct way with modern CMake I believe.

I think I should remove [this](https://github.com/openMVG/openMVG/blob/master/src/CMakeLists.txt#L552-L556). I'm not sure it is still used.

BTW, the build without unordered map is working https://travis-ci.org/openMVG/openMVG.
But trouble could certainly appears when you install and use it.

**Comment by sebastienchappuis** [comment]
Yes building works as intended but linking from outside causes crashes.

I can push a fix using target_compile_options if that is fine for you.

**Comment by sebastienchappuis** [comment]
note that OPENMVG_USE_OPENMP is not passed to the targets either.

**Comment by sebastienchappuis** [comment]
So I'm thinking of automatizing the flag passing to installed targets so that we do not have to remember to use `target_compile_options` each time we add a target or a flag. I will probably do this by giving its purpose back to the `register_definitions` macro. This would automatically pass the OPENMVG_USE_OPENMP and OPENMVG_STD_UNORDERED_MAP flags to all installed targets.

Now I could either loop over each built target using the [`BUILDSYSTEM_TARGET`](https://cmake.org/cmake/help/latest/prop_dir/BUILDSYSTEM_TARGETS.html#prop_dir:BUILDSYSTEM_TARGETS) construct, but this is only available since cmake 3.7 and we would have to update the cmake requirements of openMVG...

The other solution is to overload add_executable, add_library and add_custom_target as suggested [here](https://stackoverflow.com/a/36341753). This has the advantage of still being compatible with cmake 3.1 but it is a bit more cumbersome codewise.

Which way would you prefer, or do you see a third solution ? I will gladly implement it.

**Comment by sebastienchappuis** [comment]
hello @pmoulon, did you have time to look this up ?

I did some investigation on my side and as far as I understood the `BUILDSYSTEM_TARGET` is not as useful as i thought it would be...Basically it only lists the targets of a specific dir, so you have to first list all the directories you want to add definitions to, which is not as easy to automatize as I thought... 

It would look something like this :
```
set(dirlist openMVG third_party ... other relevant dirs)
set(all_subdirs_list "")
foreach(dir ${dirlist})
  get_directory_property(subdir_list DIRECTORY ${dir} SUBDIRECTORIES)
  list(APPEND all_subdirs_list ${subdir_list})
endforeach()

foreach(subdir ${all_subdirs_list})
  get_property(BUILDSYSTEM_TARGETS DIRECTORY ${subdir} PROPERTY BUILDSYSTEM_TARGETS)
  foreach(tget ${BUILDSYSTEM_TARGETS})
    get_target_property(target_type ${tget} TYPE)
    if (target_type STREQUAL "INTERFACE_LIBRARY")
      target_compile_definitions(${tget} INTERFACE ${OpenMVG_DEFINITIONS})
    elseif(target_type STREQUAL "STATIC_LIBRARY")
      target_compile_definitions(${tget} PUBLIC ${OpenMVG_DEFINITIONS})
    endif (target_type STREQUAL "INTERFACE_LIBRARY")
  endforeach(tget ${BUILDSYSTEM_TARGETS})
endforeach()
``` 

So I'm kind of skeptical about that approach now. Do you think it is still better than overloading the `add_library`, `add_executable` and `add_custom_target` functions ?

cheers

**Comment by sebastienchappuis** [comment]
@pmoulon thank you for your answer!

1. I did not measure anything, no. I'm just using the default Hash_Map when compiling with clang.
2. agreed :)
3. this could work, but then this does not solve the problem for the OPENMVG_USE_OPENMP definition... I think it would be better to have a solution which works for every compile option otherwise the problem could reappear later on when adding new compile options.

**Comment by pmoulon** [comment]
I think this is a similar issue as https://github.com/openMVG/openMVG/issues/1474
I set it the issue as a must fix for a future revision to be sure to take a look at it.
